### DISCUSSION TITLE
Resolved: Why don't we depend on `hex` crate

### DISCUSSION THREAD

**Discussion by tcharding** [rust-bitcoin/rust-bitcoin#1243] [General]
Title is the TL;DR 

ref: https://crates.io/crates/hex 

(I did try and search for previous discussion on this topic)

We are attempting to crate smash various crates in our stack, pulling out hex functionality into a crate seems likely. I took a look at the hex crate and except for the dev dependencies playing havoc with Rust 1.41.1 it looks pretty good (its only 700 LOC). 

Is this perhaps a good time to start using crev properly and auditing a dependency?

If we create a `bitcoin-hex` crate, since there is nothing special about hex handling in bitcoin, an outsider would almost have to judge that we suffer from NIH syndrome. 

To explore the idea I did:

- A branch removing `ToHex` and `FromHex` from `bitcoin_hashes` (just implemented `FromStr` and `ToString`, no new dependency added): https://github.com/rust-bitcoin/bitcoin_hashes/pull/183
- A branch depending on patched `bitcoin_hashes` and `hex` crate: https://github.com/rust-bitcoin/rust-bitcoin/pull/1242

An alternative would be to depend on `hex` in `bitcoin_hashes` as well?

**Comment by sanket1729** [comment]
> ref: [crates.io/crates/hex](https://crates.io/crates/hex)

Only has one `serde` dependency, which is a good thing. But that would also mean that we have the compilation time with serde.  Do you see a significant performance difference in compilation time? 

If we bring in serde in regular dependencies, then we can remove the feature.

**Reply by TheBlueMatt** [reply]
Given my recent interaction with the serde maintainer, making serde and explicit dependency seems unacceptable. Luckily it's optional in hex as well, so it wouldn't have to be here.

**Comment by apoelstra** [comment]
Our hex implementation is smaller and serves exactly our purposes -- though actually I believe we are going to shrink it even further. We've talked in the past about pulling it out of `bitcoin_hashes` and putting it into its own crate maybe with bech32 or base58 or base64[*], but adding a 3rd party dependency does not sound good.

"NIH syndrome" in Rust land is common sense. We should always default to writing our own code because almost everyone else in the ecosystem has different and conflicting goals than us (charitably speaking). The only cases we'd *not* NIH something is if there is significant complexity (no) or common traits that we need to share with the rest of the ecosostym (no).

[*] bech32 has its own crate maintained by Clark Moody somewhat separately from us, and base58 annoyingly has a dependency on bitcoin_hashes which would lead to circularity, but I think base64 and hex definitely ought to share a crate.

**Comment by TheBlueMatt** [comment]
Another way to ask this - could you imagine a C library taking on a dependency in order to encode hex? I don't think I really could, and I think that's a relevant question here - just because it's easy to add the dependency line doesn't mean we should.

**Comment by tcharding** [comment]
Thanks fellas, very informative.

> "NIH syndrome" in Rust land is common sense. We should always default to writing our own code because almost everyone else in the ecosystem has different and conflicting goals than us (charitably speaking). 

I'd like to think that there others in the world writing code trying to keep to the high standards that we are trying to attain, it would be nice to collaborate with those people. Or are there strictly technical things also (like maintaining support for an MSRV at least two years old)?

> The only cases we'd not NIH something is if there is significant complexity (no) or common traits that we need to share with the rest of the ecosostym (no).

Cool, thanks for the clarification.

**Comment by Kixunil** [comment]
I only noticed this discussion now an want to bring in a different view on dependencies in general. I'm much more open to adding dependencies because it means sharing code among more developers which means more eyes which means less bugs. The difference is small in case of simple things like `hex`, especially when there is no suitable 3rd party crate, so I'm OK with NIH there.

Even when our goals are different than the goals of the maintainers of existing crates I doubt we, the maintainers, are the only ones who have such goals. If we can find the others it'd still be better to share a crate. Not something I'll bother for hex encoding unless I come across such people randomly.

> If we create a `bitcoin-hex` crate, since there is nothing special about hex handling in bitcoin, an outsider would almost have to judge that we suffer from NIH syndrome.

`hex-compat` was suggested before, I'm now also thinking about `hex-conservative`. These names along with the docs should explain the differences. But I'd just keep it in internals.

I'm also not sure what exactly our requirements are. It'd be helpful to write them down. These candidates come to my mind:

* Maintainers are reputable and know idiomatic Rust well-enough
* Maintainers respond to bug reports quickly
* Idiomatic API, well documented, ideally stable or close to stable
* Conservative MSRV - hard requirement for mandatory dependencies
* Reasonable performance
* Good test coverage, CI, fuzzing, clippy, miri (if applicable)
* No reckless `unsafe` defined as having any of these properties: large complicated chunks of `unsafe` (whole module containing `unsafe` is an `unsafe` chunk!), (private) function not marked `unsafe` when it can UB depending on arguments (`actix-web` flamewar), code that can be written without `unsafe` while staying equally fast thanks to compiler optimizations, missing safe abstractions.
* Maintainer responds well to code quality PRs (refactors)

**Reply by apoelstra** [reply]
That's a pretty solid list. I'd add a requirement that their own dependency trees are not gratuitously large.

**Reply by Kixunil** [reply]
How about instead say that the requirements apply recursively and the total dependency tree must not contain significantly more code than what is required to implement the desired feature? (This is a bit ironic but hey, we're working on it. ;))

**Comment by tcharding** [comment]
Policy from this discussion added to `docs/` in #4854.
